net/graphsync_fetcher_test.go

```diff
@@ -0,0 +1,284 @@
+	"github.com/ipld/go-ipld-prime/traversal/selector"
+	"github.com/libp2p/go-libp2p-core/peer"
+	mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
+	"github.com/stretchr/testify/require"
+)
+
+type mockRequest struct {
+	root     ipld.Link
+	selector selector.Selector
+}
+
+type mockResponse struct {
+	responses []graphsync.ResponseProgress
+	errs      []error
+}
+
+type requestResponse struct {
+	request  mockRequest
+	response mockResponse
+}
+
+type mockableGraphsync struct {
+	stubs            []requestResponse
+	receivedRequests []mockRequest
+}
+
+func toChans(mr mockResponse) (<-chan graphsync.ResponseProgress, <-chan error) {
+	errChan := make(chan error, len(mr.errs))
+	for _, err := range mr.errs {
+		errChan <- err
+	}
+	close(errChan)
+
+	responseChan := make(chan graphsync.ResponseProgress, len(mr.responses))
+	for _, response := range mr.responses {
+		responseChan <- response
+	}
+	close(responseChan)
+
+	return responseChan, errChan
+}
+
+func (mgs *mockableGraphsync) Request(ctx context.Context, p peer.ID, root ipld.Link, selectorSpec ipld.Node) (<-chan graphsync.ResponseProgress, <-chan error) {
+	parsed, err := selector.ParseSelector(selectorSpec)
+	if err != nil {
+		return toChans(mockResponse{nil, []error{fmt.Errorf("Invalid selector")}})
+	}
+	request := mockRequest{root, parsed}
+	mgs.receivedRequests = append(mgs.receivedRequests, request)
+	for _, stub := range mgs.stubs {
+		if reflect.DeepEqual(stub.request, request) {
+			return toChans(stub.response)
+		}
+	}
+	return toChans(mockResponse{nil, []error{fmt.Errorf("Failed Request")}})
+}
+
+func makeGsResponse(path string, blockCid cid.Cid) graphsync.ResponseProgress {
+	return graphsync.ResponseProgress{
+		Path: ipld.ParsePath(path),
+		LastBlock: struct {
+			Path ipld.Path
+			Link ipld.Link
+		}{
+			Link: cidlink.Link{Cid: blockCid},
+		},
+	}
+}
+
+func TestGraphsyncFetcher(t *testing.T) {
+	tf.UnitTest(t)
+	ctx := context.Background()
+	bs := bstore.NewBlockstore(dss.MutexWrap(datastore.NewMapDatastore()))
+	bv := th.NewFakeBlockValidator()
+
+	ssb := selector.NewSelectorSpecBuilder(ipldfree.NodeBuilder())
+	layer1Selector, err := ssb.Matcher().Selector()
+	require.NoError(t, err)
+	gsSelector, err := ssb.ExploreRecursive(1, ssb.ExploreFields(func(efsb selector.ExploreFieldsSpecBuilder) {
+		efsb.Insert("parents", ssb.ExploreUnion(
+			ssb.ExploreAll(ssb.Matcher()),
+			ssb.ExploreIndex(0, ssb.ExploreRecursiveEdge()),
+		))
+	})).Selector()
+	require.NoError(t, err)
+
+	t.Run("happy path returns correct tipsets", func(t *testing.T) {
+		parentBlock := types.NewBlockForTest(nil, uint64(0))
+		block1 := types.NewBlockForTest(parentBlock, uint64(3))
+		block2 := types.NewBlockForTest(parentBlock, uint64(5))
+		block3 := types.NewBlockForTest(parentBlock, uint64(6))
+		requireBlockStorePut(t, bs, parentBlock.ToNode())
+		requireBlockStorePut(t, bs, block1.ToNode())
+		requireBlockStorePut(t, bs, block2.ToNode())
+		requireBlockStorePut(t, bs, block3.ToNode())
```

Comment:
I find this test quite confusing because the blocks are already in the store. I would expect them to be put there only as a side-effect of the graphsync request. I think I understand what's going on and why this test works, but it's very white-box, and very legitimate changes to the internals of the fetcher could break this test without changing essential functionality.

Since you have started with a mock graphsync (thanks!) let's make it better and have the mock responsible for placing blocks in the store. This will then remain a good test if we dispense with the fetcher return value and lean in to relying on the storage side-effect.